### PR TITLE
feat: regularized NDT matching

### PR DESCRIPTION
**Pull request by KYabuuchi**
## Description
A function that uses GNSS or other information to prevent the NDT slipping.
> This PR is derived from https://github.com/tier4/autoware.iv/pull/2180 (TIER IV private)

When regularization is enabled, ndt_scan_matcher subscribes to the regularization refenrence pose and assigns a regularization term to the NDT.

Regularization is disabled by default. We can toggle it from `ndt_scan_matcher.param.yaml` . 
This regularization is only available with NDT_OMP.  
When other NDT types (`PCL_GENERIC` or `PCL_MODIFIED`) are used, the regularization is ignored.

## Related links
- https://github.com/tier4/autoware.iv/pull/2180 (TIER IV private)
- https://github.com/tier4/autoware.iv/pull/2327 (TIER IV private)
- https://github.com/tier4/ndt_omp/pull/6

## Tests performed
I used the dataset from https://github.com/tier4/autoware.iv/pull/2327 (TIER IV private).
It includes two maps. One has many features and the other has fewer features.

In conventional NDT, estimated position slips on a straight section in the feature-less map, as shown in the figure below.

The color of the trajectory indicates the error from the localization result with the feature-rich map. 
The maximum error was **13m**.

When regularization was enabled, the estimation results improved.
> I set  `/sensing/gnss/pose_with_covariance` for the `regularization_pose` and `0.01` for the `regularization_scale_factor`.

The maximum error was **1.3m**. The estimated position is completely recovered at the end of the bridge.

## Notes for reviewers
This is my first PR :smile: 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: localization/ndt_scan_matcher/README.md [hunk 1]
```diff
@@ -59,3 +64,11 @@ There are two main functions in this package:
 | `converged_param_transform_probability` | double | Threshold for deciding whether to trust the estimation result                                   |
 | `omp_neighborhood_search_method`        | int    | neighborhood search method in OMP (0=KDTREE, 1=DIRECT26, 2=DIRECT7, 3=DIRECT1)                  |
 | `omp_num_threads`                       | int    | Number of threads used for parallel computing                                                   |
+
+### Regularization Parameters
+
+Regularization is available only with NDT_OMP.
+| Name | Type | Description |
+| --------------------------------------- | ------ | ----------------------------------------------------------------------- |
+| `regularization_enabled` | bool | Flag to add regulalization term to NDT optimization (FALSE by default) |
```
Comment (kenji-miyake) [inline localization/ndt_scan_matcher/README.md:73]:
```suggestion
| `regularization_enabled` | bool | Flag to add regularization term to NDT optimization (FALSE by default) |
```

**Comment by kenji-miyake** [comment]
@KYabuuchi Thank you for your first PR!
Could you set an appropriate reviewer following the pull request guidelines? :bow: 
https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

File: localization/ndt_scan_matcher/README.md (hunk 1 above)
Comment (KYabuuchi) [reply localization/ndt_scan_matcher/README.md]: Thenks for your suggestion.  I fixed it in [23ec0c9](https://github.com/autowarefoundation/autoware.universe/pull/1006/commits/23ec0c938c2551293602d6a97237a85dc66634d9)

File: localization/ndt_scan_matcher/README.md [hunk 2]
```diff
@@ -59,3 +64,11 @@ There are two main functions in this package:
...
 | `omp_neighborhood_search_method`        | int    | neighborhood search method in OMP (0=KDTREE, 1=DIRECT26, 2=DIRECT7, 3=DIRECT1)                  |
 | `omp_num_threads`                       | int    | Number of threads used for parallel computing                                                   |
+
+### Regularization Parameters
+
+Regularization is available only with NDT_OMP.
+| Name | Type | Description |
+| --------------------------------------- | ------ | ----------------------------------------------------------------------- |
+| `regularization_enabled` | bool | Flag to add regularization term to NDT optimization (FALSE by default) |
+| `regularization_scale_factor` | double | Coefficient of the regularization term |
```
Comment (yukkysaito) [inline localization/ndt_scan_matcher/README.md:74]: pls apply formatter

**Review by yukkysaito — APPROVED** [review]
LGTM
Please get @YamatoAndo  approval as well.

File: launch/tier4_localization_launch/config/ndt_scan_matcher.param.yaml [hunk 3]
```diff
@@ -62,3 +62,9 @@
         0.0,   0.0,   0.0,   0.0,      0.000625, 0.0,
         0.0,   0.0,   0.0,   0.0,      0.0,      0.000625,
       ]
+
+    # GNSS Regularization switch
```
Comment (hiro-ya-iv) [inline launch/tier4_localization_launch/config/ndt_scan_matcher.param.yaml:66]: @KYabuuchi 
The poses provided for regularization are not limited to GNSS. 
I don't think you need to add the word "GNSS".

Comment (KYabuuchi) [reply launch/tier4_localization_launch/config/ndt_scan_matcher.param.yaml]: Thanks for your reviewing!! I removed the word in [5ef69cf](https://github.com/autowarefoundation/autoware.universe/pull/1006/commits/5ef69cffe40ee8df703a1749542fdf5526b50684)

**Comment by KYabuuchi** [comment]
I'm writing the documentation for this feature in the README :muscle:

File: localization/ndt_scan_matcher/README.md [hunk 4]
```diff
@@ -59,3 +64,90 @@ There are two main functions in this package:
...
+            0&0&0
+        \end{pmatrix}
+        (\mathbf{t}-\mathbf{t_{base}})
+        \right\|_{2}^2
+\end{align}
+$$
+
+, where t_base is base position measured by GNSS or other means.
+NDT(R,t) stands for the pure NDT cost function.
+The regularization term shifts the optimal solution to the base position.
```
Comment (hiro-ya-iv) [inline localization/ndt_scan_matcher/README.md:91]: @KYabuuchi 
Thank you for adding to the README.
I think it's almost correct, but I don't think there is enough explanation for this part.
The regularization term shifts the optimal solution towards the base position only in the longitudinal axis of vehicle.
[ndt_omp source code](https://github.com/tier4/ndt_omp/blob/35ea99ad594cbc72233d6269132db1cf7d45bf49/include/pclomp/ndt_omp_impl.hpp#L342)

Comment (KYabuuchi) [reply localization/ndt_scan_matcher/README.md:91]: Thanks! I was confused. I will fix it later :runner:

Comment (KYabuuchi) [reply localization/ndt_scan_matcher/README.md]: @hiro-ya-iv I have corrected the issues you pointed out.
Please check again in your free time :bow:

Comment (hiro-ya-iv) [reply localization/ndt_scan_matcher/README.md]: LGTM

**Comment by yukkysaito** [comment]
@hiro-ya-iv @YamatoAndo ping

**Review by hiro-ya-iv — APPROVED** [review]

**Pull request merged**